pbsmetrics/pbsmetrics.go

```diff
@@ -0,0 +1,350 @@
+	NoCookieMeter       metrics.Meter
+	SafariRequestMeter  metrics.Meter
+	SafariNoCookieMeter metrics.Meter
+	ErrorMeter          metrics.Meter
+	InvalidMeter        metrics.Meter
+	RequestTimer        metrics.Timer
+	CookieSyncMeter     metrics.Meter
+	// Metrics for OpenRTB requests specifically. So we can track what % of RequestsMeter are OpenRTB
+	// and know when legacy requests have been abandoned.
+	ORTBRequestMeter metrics.Meter
+
+	AdapterMetrics map[openrtb_ext.BidderName]*AdapterMetrics
+
+	exchanges []openrtb_ext.BidderName
+}
+
+type AdapterMetrics struct {
+	NoCookieMeter     metrics.Meter
+	ErrorMeter        metrics.Meter
+	NoBidMeter        metrics.Meter
+	TimeoutMeter      metrics.Meter
+	RequestMeter      metrics.Meter
+	RequestTimer      metrics.Timer
+	PriceHistogram    metrics.Histogram
+	BidsReceivedMeter metrics.Meter
+}
+
+// Create a new Metrics object with all blank metrics object. This may also be useful for
+// testing routines to ensure that no metrics are written anywhere.
+
+func NewBlankMetrics(registry metrics.Registry, exchanges []openrtb_ext.BidderName) *Metrics {
+	newMetrics := &Metrics{
+		metricsRegistry:     registry,
+		RequestMeter:        blankMeter(0),
+		AppRequestMeter:     blankMeter(0),
+		NoCookieMeter:       blankMeter(0),
+		SafariRequestMeter:  blankMeter(0),
+		SafariNoCookieMeter: blankMeter(0),
+		ErrorMeter:          blankMeter(0),
+		InvalidMeter:        blankMeter(0),
+		RequestTimer:        blankTimer(0),
+		CookieSyncMeter:     blankMeter(0),
+		ORTBRequestMeter:    blankMeter(0),
+
+		AdapterMetrics: make(map[openrtb_ext.BidderName]*AdapterMetrics, len(exchanges)),
+
+		exchanges: exchanges,
+	}
+	for _, a := range exchanges {
+		newMetrics.AdapterMetrics[a] = makeBlankAdapterMetrics(registry, a)
+	}
+
+	return newMetrics
+}
+
+// Create a new Metrics object with needed metrics defined. In time we may develop to the point
+// where Metrics contains all the metrics we might want to record, and then we build the actual
+// metrics object to contain only the metrics we are interested in. This would allow for debug
+// mode metrics. The code would allways try to record the metrics, but effectively noop if we are
+// using a blank meter/timer.
+
+func NewMetrics(registry metrics.Registry, exchanges []openrtb_ext.BidderName) *Metrics {
+	newMetrics := NewBlankMetrics(registry, exchanges)
+	newMetrics.RequestMeter = metrics.GetOrRegisterMeter("requests", registry)
+	newMetrics.SafariRequestMeter = metrics.GetOrRegisterMeter("safari_requests", registry)
+	newMetrics.ErrorMeter = metrics.GetOrRegisterMeter("error_requests", registry)
+	newMetrics.NoCookieMeter = metrics.GetOrRegisterMeter("no_cookie_requests", registry)
+	newMetrics.AppRequestMeter = metrics.GetOrRegisterMeter("app_requests", registry)
+	newMetrics.SafariNoCookieMeter = metrics.GetOrRegisterMeter("safari_no_cookie_requests", registry)
+	newMetrics.RequestTimer = metrics.GetOrRegisterTimer("request_time", registry)
```

Comment:
Can we rename these so that they dump to new places?

Reason being: traffic will shift over gradually. Ideally we catch any issues in OpenRTB early--say, when only 1% of the traffic is using it.

It would be very helpful if we could compare some of these side-by-side. For example, info about bids per adapter would help us find any bugs in the upgraded APN or Rubi adapters. The timers would help us compare performance. etc